pkg/services/provisioning/datasources/types.go

```diff
@@ -116,6 +117,19 @@ func (cfg *DatasourcesAsConfigV1) mapToDatasourceFromConfig(apiVersion int64) *D
 	}
 
 	for _, ds := range cfg.Datasources {
+
+		// To provide better security, core datasources store passwords in secureJsonData but that was not always the
+		// case that is why provisioning files can use password field directly. This normalizes it so both password
+		// field and secureJsonData can be used in config. For non core datasources they have to opt in explicitly to
+		// use secureJsonData.
+		if plugins.IsCoreDatasource(ds.Type) {
```

Comment:
Special behaviour for core datasources. Not nice but keeps the format backward compatible.

Reply:
Is this special behaviour really needed given the fallbacks used in `DecryptedBasicAuthPassword` and `DecryptedPassword`?

Reply:
And also shouldn't this logic be applied when creating/updating new/existing datasources through api as well?

Reply:
> Is this special behaviour really needed given the fallbacks used in DecryptedBasicAuthPassword and DecryptedPassword?

Not needed but it would mean that you will have passwords plaintext until you manually change your provisioning config. Also as we already have provisioning which use both `password` field and `secureJsonData: { password }` depending on the data source and I am not sure if it makes sense to break these. Like we could create a new version where you would just use `password` that would be automatically encrypted and store in `secureJsonData`.

> And also shouldn't this logic be applied when creating/updating new/existing datasources through api as well?

Yes I did not thought about that to be honest but probably should be in sync to some extent.

Reply:
> Not needed but it would mean that you will have passwords plaintext until you manually change your provisioning config.

Wouldn't the database migration take care of that? Or is the migration only updating certain datasource types and not all?

Reply:
Migration will only change what ever is in DB right now so I wanted to make sure this holds:
`old datasource + migration = new datasource` if that makes sense, and that mean we need to migrate all datasources in the DB and have a new provisioning.
But maybe this is not needed?
This all is done only for core datasources so the premise of this PR is that you can still use both encrypted or plaintext storage for password. Maybe it would be enough to just say `"change your provisioning files and reset your passwords after the update if you want passwords to be encrypted"` in the change log and so the migration would not be needed?

Reply:
Ah okay my bad! Thought we we're migrating "all" datasources. In that case I think this is okay during migration phase. Think we should question when we're feeling sort of done with this PR whether it would make sense to migrate all datasources or not given that the code may be easier to reason about and/or decide when we migrate all the other datasources.